[MiHome] Secondary Xiaomi gateway bridge not initialised

<!-- Provide a general summary of the issue in the *Title* above -->
<!-- If the issue is related to a binding, please include its short name in -->
<!-- square brackets in the title - Example: "[astro] My issue..." -->

<!-- Important: Please contact the openHAB community forum for questions or -->
<!-- for configuration and usage guidance: https://community.openhab.org -->

<!-- Feel free to delete any comment lines in the template (starting with "<!--") -->

## Expected Behavior

Being able to use 2 xiaomi gateways.

## Current Behavior
After upgrading from 2.5.3 to 2.5.4 my second xiaomi gateway doens't work anymore.
I never had any problems before.
Downgrading mihome binding to 2.5.3 solves the problem

Log:

2020-04-19 20:57:10.516 [hingStatusInfoChangedEvent] - 'mihome:bridge:XXXXX' changed from UNINITIALIZED to INITIALIZING

==> /var/log/openhab2/openhab.log <==

2020-04-19 20:57:11.074 [ERROR] [nternal.DiscoveryServiceRegistryImpl] - Cannot trigger scan for thing types '[mihome:sensor_ht, mihome:86sw2, mihome:sensor_magnet_aq2, mihome:86sw1, mihome:basic, mihome:sensor_plug, mihome:ctrl_ln1, mihome:ctrl_ln2, mihome:sensor_magnet, mihome:gateway, mihome:curtain, mihome:smoke, mihome:sensor_motion_aq2, mihome:sensor_switch_aq2, mihome:sensor_weather_v1, mihome:sensor_wleak_aq1, mihome:sensor_switch, mihome:sensor_motion, mihome:sensor_vibration, mihome:sensor_cube, mihome:sensor_lock_aq1, mihome:natgas, mihome:ctrl_neutral1, mihome:ctrl_neutral2]' on 'XiaomiItemDiscoveryService'!

java.lang.NullPointerException: null

	at org.openhab.binding.mihome.internal.socket.XiaomiSocket.sendMessage(XiaomiSocket.java:152) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendMessageToBridge(XiaomiBridgeHandler.java:272) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendCommandToBridge(XiaomiBridgeHandler.java:303) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendCommandToBridge(XiaomiBridgeHandler.java:276) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.readDeviceList(XiaomiBridgeHandler.java:145) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.discoverItems(XiaomiBridgeHandler.java:392) ~[?:?]

	at org.openhab.binding.mihome.internal.discovery.XiaomiItemDiscoveryService.startScan(XiaomiItemDiscoveryService.java:68) ~[?:?]

	at org.eclipse.smarthome.config.discovery.AbstractDiscoveryService.startScan(AbstractDiscoveryService.java:209) ~[bundleFile:?]

	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:382) [bundleFile:?]

	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScans(DiscoveryServiceRegistryImpl.java:358) [bundleFile:?]

	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:216) [bundleFile:?]

	at org.eclipse.smarthome.io.rest.core.internal.discovery.DiscoveryResource.scan(DiscoveryResource.java:97) [bundleFile:?]

	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222]

	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222]

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222]

	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222]

	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) [bundleFile:?]

	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) [bundleFile:?]

	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) [bundleFile:?]

	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160) [bundleFile:?]

	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) [bundleFile:?]

	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) [bundleFile:?]

	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) [bundleFile:?]

	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) [bundleFile:?]

	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) [bundleFile:?]

	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) [bundleFile:?]

	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) [bundleFile:?]

	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) [bundleFile:?]

	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) [bundleFile:?]

	at org.glassfish.jersey.internal.Errors.process(Errors.java:267) [bundleFile:?]

	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) [bundleFile:?]

	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) [bundleFile:?]

	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) [bundleFile:?]

	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473) [bundleFile:?]

	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427) [bundleFile:?]

	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388) [bundleFile:?]

	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341) [bundleFile:?]

	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228) [bundleFile:?]

	at com.eclipsesource.jaxrs.publisher.internal.ServletContainerBridge.service(ServletContainerBridge.java:76) [bundleFile:?]

	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:544) [bundleFile:9.4.20.v20190813]

	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [bundleFile:?]

	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:536) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307) [bundleFile:9.4.20.v20190813]

	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:293) [bundleFile:?]

	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [bundleFile:9.4.20.v20190813]

	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [bundleFile:?]

	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.Server.handle(Server.java:494) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:374) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782) [bundleFile:9.4.20.v20190813]

	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918) [bundleFile:9.4.20.v20190813]

	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]

==> /var/log/openhab2/events.log <==

2020-04-19 20:57:11.246 [vent.ItemStateChangedEvent] - Gateway_woonkamer_Illumination changed from 348 to 346

==> /var/log/openhab2/openhab.log <==

2020-04-19 20:57:11.547 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 

java.lang.NullPointerException: null

	at org.openhab.binding.mihome.internal.socket.XiaomiSocket.sendMessage(XiaomiSocket.java:152) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendMessageToBridge(XiaomiBridgeHandler.java:272) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendCommandToBridge(XiaomiBridgeHandler.java:303) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.sendCommandToBridge(XiaomiBridgeHandler.java:276) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.readDeviceList(XiaomiBridgeHandler.java:145) ~[?:?]

	at org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler.lambda$0(XiaomiBridgeHandler.java:139) ~[?:?]

	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_222]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_222]

	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_222]

	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_222]

	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]




and after adding the developer key in the thing settings:

2020-04-19 20:59:11.670 [vent.ConfigStatusInfoEvent] - ConfigStatusInfo [configStatusMessages=[]]

2020-04-19 20:59:11.675 [vent.ConfigStatusInfoEvent] - ConfigStatusInfo [configStatusMessages=[]]

## Possible Solution

Downgrading mihome binding to 2.5.3 solves the problem.



## Steps to Reproduce (for Bugs)

Add a seconday xiaomi gateway of type: lumi.gateway.v3

## Context
I use a second gateway on the top floor of my house to extend the range of the zigbee network.
Sensors on the top floor become unstable/unreachable when i use only one gateway.

## Your Environment
Raspberry Pi 3B+ with openhabian 2.5.4.

Bindings in use:
airquality, astro , hue , ntp , mihome , openweathermap , network , sonyaudio , dsmr , miio , http1, mqtt, gpstracker, mail, telegram, chromecast, hpprinter

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/discovery/XiaomiBridgeDiscoveryService.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/discovery/XiaomiBridgeDiscoveryService.java
@@ -19,6 +19,7 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 
+import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
 import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
 import org.eclipse.smarthome.config.discovery.DiscoveryService;
@@ -38,6 +39,7 @@ import com.google.gson.JsonObject;
  * @author Patrick Boos - Initial contribution
  * @author Kuba Wolanin - logger fixes
  */
+@NonNullByDefault
 @Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.mihome")
 public class XiaomiBridgeDiscoveryService extends AbstractDiscoveryService implements XiaomiSocketListener {
 
@@ -45,7 +47,7 @@ public class XiaomiBridgeDiscoveryService extends AbstractDiscoveryService imple
     private static final int DISCOVERY_TIMEOUT_SEC = 30;
 
     private final Logger logger = LoggerFactory.getLogger(XiaomiBridgeDiscoveryService.class);
-    private XiaomiDiscoverySocket socket;
+    private final XiaomiDiscoverySocket socket = new XiaomiDiscoverySocket("discovery");
 
     public XiaomiBridgeDiscoveryService() {
         super(SUPPORTED_THING_TYPES, DISCOVERY_TIMEOUT_SEC, false);
@@ -53,8 +55,7 @@ public class XiaomiBridgeDiscoveryService extends AbstractDiscoveryService imple
 
     @Override
     protected void startScan() {
-        socket = (socket == null) ? new XiaomiDiscoverySocket() : socket;
-        socket.intialize();
+        socket.initialize();
         logger.debug("Start scan for bridges");
         socket.registerListener(this);
         discoverGateways();
@@ -65,17 +66,13 @@ public class XiaomiBridgeDiscoveryService extends AbstractDiscoveryService imple
         super.stopScan();
         logger.debug("Stop scan");
         removeOlderResults(getTimestampOfLastScan());
-        if (socket != null) {
-            socket.unregisterListener(this);
-        }
+        socket.unregisterListener(this);
     }
 
     @Override
     public void deactivate() {
         super.deactivate();
-        if (socket != null) {
-            socket.unregisterListener(this);
-        }
+        socket.unregisterListener(this);
     }
 
     @Override

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/handler/XiaomiBridgeHandler.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/handler/XiaomiBridgeHandler.java
@@ -131,8 +131,8 @@ public class XiaomiBridgeHandler extends ConfigStatusBridgeHandler implements Xi
             return;
         }
         logger.debug("Init socket on Port: {}", port);
-        socket = new XiaomiBridgeSocket(port);
-        socket.intialize();
+        socket = new XiaomiBridgeSocket(port, getThing().getUID().getId());
+        socket.initialize();
         socket.registerListener(this);
 
         scheduler.schedule(() -> {

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiBridgeSocket.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiBridgeSocket.java
@@ -13,10 +13,10 @@
 package org.openhab.binding.mihome.internal.socket;
 
 import java.io.IOException;
-import java.net.DatagramSocket;
 import java.net.InetAddress;
 import java.net.MulticastSocket;
 
+import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -26,35 +26,37 @@ import org.slf4j.LoggerFactory;
  * @author Dieter Schmidt - Initial contribution
  *
  */
+@NonNullByDefault
 public class XiaomiBridgeSocket extends XiaomiSocket {
 
     private final Logger logger = LoggerFactory.getLogger(XiaomiBridgeSocket.class);
 
-    public XiaomiBridgeSocket(int port) {
-        super(port);
+    public XiaomiBridgeSocket(int port, String owner) {
+        super(port, owner);
     }
 
     /**
      * Sets up the {@link XiaomiBridgeSocket}.
      *
      * Connects the socket to the specific multicast address and port.
-     * Starts the {@link ReceiverThread} for the socket.
      */
     @Override
-    synchronized DatagramSocket setupSocket() {
-        DatagramSocket openSocket = getOpenSockets().get(getPort());
-        if (openSocket != null) {
-            return openSocket;
+    protected synchronized void setupSocket() {
+        MulticastSocket socket = (MulticastSocket) getSocket();
+        if (socket != null) {
+            logger.debug("Socket already setup");
+            return;
         }
+
         try {
             logger.debug("Setup socket");
-            setSocket(new MulticastSocket(getPort())); // must bind receive side
-            ((MulticastSocket) getSocket()).joinGroup(InetAddress.getByName(MCAST_ADDR));
-            logger.debug("Initialized socket to {}:{} on {}:{}", getSocket().getRemoteSocketAddress(),
-                    getSocket().getPort(), getSocket().getLocalAddress(), getSocket().getLocalPort());
+            socket = new MulticastSocket(getPort());
+            setSocket(socket); // must bind receive side
+            socket.joinGroup(InetAddress.getByName(MCAST_ADDR));
+            logger.debug("Initialized socket to {}:{} on {}:{}", socket.getRemoteSocketAddress(), socket.getPort(),
+                    socket.getLocalAddress(), socket.getLocalPort());
         } catch (IOException e) {
             logger.error("Setup socket error", e);
         }
-        return getSocket();
     }
 }

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiDiscoverySocket.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiDiscoverySocket.java
@@ -17,6 +17,7 @@ import java.net.DatagramSocket;
 import java.net.InetAddress;
 import java.net.UnknownHostException;
 
+import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -26,35 +27,35 @@ import org.slf4j.LoggerFactory;
  * @author Dieter Schmidt - Initial contribution
  *
  */
+@NonNullByDefault
 public class XiaomiDiscoverySocket extends XiaomiSocket {
 
     private static final int MCAST_PORT = 4321;
 
     private final Logger logger = LoggerFactory.getLogger(XiaomiDiscoverySocket.class);
 
-    public XiaomiDiscoverySocket() {
-        super();
+    public XiaomiDiscoverySocket(String owner) {
+        super(owner);
     }
 
     /**
      * Sets up the {@link XiaomiDiscoverySocket}.
      *
      * Connects the socket to the specific multicast address and port.
-     * Starts the {@link ReceiverThread} for the socket.
      */
     @Override
-    DatagramSocket setupSocket() {
+    protected void setupSocket() {
         synchronized (XiaomiDiscoverySocket.class) {
             try {
                 logger.debug("Setup discovery socket");
-                setSocket(new DatagramSocket(0));
-                logger.debug("Initialized socket to {}:{} on {}:{}", getSocket().getInetAddress(),
-                        getSocket().getPort(), getSocket().getLocalAddress(), getSocket().getLocalPort());
+                DatagramSocket socket = new DatagramSocket(0);
+                setSocket(socket);
+                logger.debug("Initialized socket to {}:{} on {}:{}", socket.getInetAddress(), socket.getPort(),
+                        socket.getLocalAddress(), socket.getLocalPort());
             } catch (IOException e) {
                 logger.error("Setup socket error", e);
             }
         }
-        return getSocket();
     }
 
     /**

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocket.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocket.java
@@ -17,11 +17,11 @@ import java.net.DatagramPacket;
 import java.net.DatagramSocket;
 import java.net.InetAddress;
 import java.nio.charset.StandardCharsets;
-import java.util.List;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
 
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
 import org.openhab.binding.mihome.internal.discovery.XiaomiBridgeDiscoveryService;
 import org.openhab.binding.mihome.internal.handler.XiaomiBridgeHandler;
 import org.slf4j.Logger;
@@ -38,71 +38,66 @@ import com.google.gson.JsonParser;
  * @author Dieter Schmidt - JavaDoc, refactored, reviewed
  *
  */
+@NonNullByDefault
 public abstract class XiaomiSocket {
 
     static final String MCAST_ADDR = "224.0.0.50";
+
     private static final int BUFFER_LENGTH = 1024;
-    private DatagramPacket datagramPacket = new DatagramPacket(new byte[BUFFER_LENGTH], BUFFER_LENGTH);
-
-    private List<XiaomiSocketListener> listeners = new CopyOnWriteArrayList<>();
-
     private static final JsonParser PARSER = new JsonParser();
 
     private final Logger logger = LoggerFactory.getLogger(XiaomiSocket.class);
 
-    private static Map<Integer, DatagramSocket> openSockets = new ConcurrentHashMap<>();
+    private final DatagramPacket datagramPacket = new DatagramPacket(new byte[BUFFER_LENGTH], BUFFER_LENGTH);
+    private final Set<XiaomiSocketListener> listeners = new CopyOnWriteArraySet<>();
 
-    private int port;
-    private DatagramSocket socket;
-    private Thread socketReceiveThread;
+    private final int port;
+    private @Nullable DatagramSocket socket;
+    private final Thread socketReceiveThread = new Thread(this::receiveData);
 
     /**
      * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a random port
      *
+     * @param owner identifies the socket owner
      */
-    public XiaomiSocket() {
-        this(0);
+    public XiaomiSocket(String owner) {
+        this(0, owner);
     }
 
     /**
      * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a specific port
      *
-     * @param port - the socket will be bound to this port
+     * @param port the socket will be bound to this port
+     * @param owner identifies the socket owner
      */
-    public XiaomiSocket(int port) {
+    public XiaomiSocket(int port, String owner) {
         this.port = port;
+        socketReceiveThread.setName("XiaomiSocketReceiveThread(" + port + ", " + owner + ")");
     }
 
-    public void intialize() {
+    public void initialize() {
         setupSocket();
-        runReceiveThread();
-    }
-
-    protected void runReceiveThread() {
-        socketReceiveThread = new ReceiverThread();
-        socketReceiveThread.start();
-        if (getSocket() != null) {
-            getOpenSockets().put(getSocket().getLocalPort(), getSocket());
-            logger.debug("There are {} open sockets: {}", getOpenSockets().size(), getOpenSockets());
+        if (!socketReceiveThread.isAlive()) {
+            logger.trace("Starting receiver thread {}", socketReceiveThread);
+            socketReceiveThread.start();
         }
     }
 
-    abstract DatagramSocket setupSocket();
+    protected abstract void setupSocket();
 
     /**
      * Interrupts the {@link ReceiverThread} and closes the {@link XiaomiSocket}.
      */
     private void closeSocket() {
         synchronized (XiaomiSocket.class) {
-            if (socketReceiveThread != null) {
-                logger.debug("Interrupting Thread {}", socketReceiveThread);
-                socketReceiveThread.interrupt();
-            }
-            if (getSocket() != null) {
-                logger.debug("Closing socket {}", getSocket());
-                openSockets.remove(getSocket().getLocalPort());
-                getSocket().close();
-                setSocket(null);
+            logger.debug("Interrupting receiver thread {}", socketReceiveThread);
+            socketReceiveThread.interrupt();
+
+            DatagramSocket socket = this.socket;
+            if (socket != null) {
+                logger.debug("Closing socket {}", socket);
+                socket.close();
+                this.socket = null;
             }
         }
     }
@@ -111,15 +106,16 @@ public abstract class XiaomiSocket {
      * Registers a {@link XiaomiSocketListener} to be called back, when data is received.
      * If no {@link XiaomiSocket} exists, when the method is called, it is being set up.
      *
-     * @param listener - {@link XiaomiSocketListener} to be called back
+     * @param listener {@link XiaomiSocketListener} to be called back
      */
     public synchronized void registerListener(XiaomiSocketListener listener) {
-        if (!getListeners().contains(listener)) {
-            logger.trace("Adding socket listener {}", listener);
-            getListeners().add(listener);
+        if (listeners.add(listener)) {
+            logger.trace("Added socket listener {}", listener);
         }
-        if (getSocket() == null) {
-            intialize();
+
+        DatagramSocket socket = this.socket;
+        if (socket == null) {
+            initialize();
         }
     }
 
@@ -127,12 +123,14 @@ public abstract class XiaomiSocket {
      * Unregisters a {@link XiaomiSocketListener}. If there are no listeners left,
      * the {@link XiaomiSocket} is being closed.
      *
-     * @param listener - {@link XiaomiSocketListener} to be unregistered
+     * @param listener {@link XiaomiSocketListener} to be unregistered
      */
     public synchronized void unregisterListener(XiaomiSocketListener listener) {
-        getListeners().remove(listener);
+        if (listeners.remove(listener)) {
+            logger.trace("Removed socket listener {}", listener);
+        }
 
-        if (getListeners().isEmpty()) {
+        if (listeners.isEmpty()) {
             closeSocket();
         }
     }
@@ -140,36 +138,35 @@ public abstract class XiaomiSocket {
     /**
      * Sends a message through the {@link XiaomiSocket} to a specific address and port
      *
-     * @param message - Message to be sent
-     * @param address - Address, to which the message shall be sent
-     * @param port - Port, through which the message shall be sent
+     * @param message the message to be sent
+     * @param address the message destination address
+     * @param port the message destination port
      */
     public void sendMessage(String message, InetAddress address, int port) {
+        DatagramSocket socket = this.socket;
+        if (socket == null) {
+            logger.error("Error while sending message (socket is null)");
+            return;
+        }
+
         try {
             byte[] sendData = message.getBytes(StandardCharsets.UTF_8);
             DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, address, port);
             logger.trace("Sending message: {} to {}:{}", message, address, port);
-            getSocket().send(sendPacket);
+            socket.send(sendPacket);
         } catch (IOException e) {
-            logger.error("Sending error", e);
+            logger.error("Error while sending message", e);
         }
     }
 
     /**
-     * @return - the port number of this {@link XiaomiSocket}
+     * @return the port number of this {@link XiaomiSocket}
      */
     public int getPort() {
         return port;
     }
 
-    /**
-     * @return - a list of already open sockets
-     */
-    public static Map<Integer, DatagramSocket> getOpenSockets() {
-        return openSockets;
-    }
-
-    protected DatagramSocket getSocket() {
+    protected @Nullable DatagramSocket getSocket() {
         return socket;
     }
 
@@ -177,71 +174,58 @@ public abstract class XiaomiSocket {
         this.socket = socket;
     }
 
-    protected List<XiaomiSocketListener> getListeners() {
-        return listeners;
+    /**
+     * This method is the main method of the receiver thread for the {@link XiaomiBridgeSocket}.
+     * If the socket has data, it parses the data to a json object and calls all
+     * {@link XiaomiSocketListener} and passes the data to them.
+     */
+    private void receiveData() {
+        DatagramSocket socket = this.socket;
+        if (socket == null) {
+            logger.error("Failed to receive data (socket is null)");
+            return;
+        }
+
+        Thread currentThread = Thread.currentThread();
+        int localPort = socket.getLocalPort();
+
+        try {
+            while (!currentThread.isInterrupted()) {
+                logger.trace("Thread {} waiting for data on port {}", currentThread.getName(), localPort);
+                socket.receive(datagramPacket);
+                InetAddress address = datagramPacket.getAddress();
+                logger.debug("Received Datagram from {}:{} on port {}", address.getHostAddress(),
+                        datagramPacket.getPort(), localPort);
+                String sentence = new String(datagramPacket.getData(), 0, datagramPacket.getLength());
+                JsonObject message = PARSER.parse(sentence).getAsJsonObject();
+                notifyListeners(message, address);
+                logger.trace("Data received and notified {} listeners", listeners.size());
+            }
+        } catch (IOException e) {
+            if (!currentThread.isInterrupted()) {
+                logger.error("Error while receiving", e);
+            } else {
+                logger.trace("Receiver thread was interrupted");
+            }
+        }
+        logger.debug("Receiver thread ended");
     }
 
     /**
-     * The thread, which waits for data on the {@link XiaomiSocket} and handles it, when received
+     * Notifies all {@link XiaomiSocketListener} on the parent {@link XiaomiSocket}. First checks for any matching
+     * {@link XiaomiBridgeHandler}, before passing to any {@link XiaomiBridgeDiscoveryService}.
      *
-     * @author Patrick Boos - Initial contribution
-     * @author Dieter Schmidt - comments and synchronized block for callback instead of copy
-     *
+     * @param message the data message as {@link JsonObject}
+     * @param address the address from which the message was received
      */
-    private class ReceiverThread extends Thread {
-        @Override
-        public void run() {
-            logger.trace("Staring reveicer thread for socket on port {}", getSocket().getLocalPort());
-            receiveData(getSocket(), datagramPacket);
-        }
-
-        /**
-         * This method is the main method of the {@link ReceiverThread} for the {@link XiaomiBridgeSocket}.
-         * If the socket has data, it parses the data to a json object and calls all
-         * {@link XiaomiSocketListener} and passes the data to them.
-         *
-         * @param socket - The multicast socket to listen to
-         * @param dgram - The datagram to receive
-         */
-        private void receiveData(DatagramSocket socket, DatagramPacket dgram) {
-            try {
-                while (true) {
-                    logger.trace("Thread {} waiting for data on port {}", this, socket.getLocalPort());
-                    socket.receive(dgram);
-                    InetAddress address = dgram.getAddress();
-                    logger.debug("Received Datagram from {}:{} on Port {}", address.getHostAddress(), dgram.getPort(),
-                            socket.getLocalPort());
-                    String sentence = new String(dgram.getData(), 0, dgram.getLength());
-                    JsonObject message = PARSER.parse(sentence).getAsJsonObject();
-                    notifyAll(getListeners(), message, address);
-                    logger.trace("Data received and notified {} listeners", getListeners().size());
-                }
-            } catch (IOException e) {
-                if (!isInterrupted()) {
-                    logger.error("Error while receiving", e);
-                } else {
-                    logger.trace("Receiver thread was interrupted");
-                }
-            }
-            logger.debug("Receiver thread ended");
-        }
-
-        /**
-         * Notifies all {@link XiaomiSocketListener} on the parent {@link XiaomiSocket}. First checks for any matching
-         * {@link XiaomiBridgeHandler}, before passing to any {@link XiaomiBridgeDiscoveryService}.
-         *
-         * @param listeners - a list of all {@link XiaomiSocketListener} to notify
-         * @param message - the data message as {@link JsonObject}
-         */
-        synchronized void notifyAll(List<XiaomiSocketListener> listeners, JsonObject message, InetAddress address) {
-            for (XiaomiSocketListener listener : listeners) {
-                if (listener instanceof XiaomiBridgeHandler) {
-                    if (((XiaomiBridgeHandler) listener).getHost().equals(address)) {
-                        listener.onDataReceived(message);
-                    }
-                } else if (listener instanceof XiaomiBridgeDiscoveryService) {
+    private void notifyListeners(JsonObject message, InetAddress address) {
+        for (XiaomiSocketListener listener : listeners) {
+            if (listener instanceof XiaomiBridgeHandler) {
+                if (((XiaomiBridgeHandler) listener).getHost().equals(address)) {
                     listener.onDataReceived(message);
                 }
+            } else if (listener instanceof XiaomiBridgeDiscoveryService) {
+                listener.onDataReceived(message);
             }
         }
     }

--- a/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocketListener.java
+++ b/bundles/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocketListener.java
@@ -12,6 +12,8 @@
  */
 package org.openhab.binding.mihome.internal.socket;
 
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
 import com.google.gson.JsonObject;
 
 /**
@@ -20,6 +22,7 @@ import com.google.gson.JsonObject;
  *
  * @author Patrick Boos - Initial contribution
  */
+@NonNullByDefault
 public interface XiaomiSocketListener {
     /**
      * Callback method for the {@link XiaomiSocketListener}

